### PR TITLE
feat(celery) record queue duration

### PR DESCRIPTION
**Pull request by scefali**
This PR adds a new option called `record_timing` which will record how long a task is in the queue before it is picked up. Eventually, we will remove the option and make recording the time the default.

### CODE REVIEW

File: src/sentry/tasks/base.py [hunk 1]
```diff
@@ -94,16 +95,27 @@ def instrumented_task(name, stat_suffix=None, silo_mode=None, **kwargs):
...
             transaction_id = kwargs.pop("__transaction_id", None)
+            start_time = kwargs.pop("__start_time", None)
 
             key = "jobs.duration"
             if stat_suffix:
                 instance = f"{name}.{stat_suffix(*args, **kwargs)}"
             else:
                 instance = name
 
+            if start_time and record_timing:
```
[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/tasks/base.py:110]: @NisanthanNanthakumar to be clear, this is _not_ to get the duration of the actual job itself, only how long it sits in the celery queue. Also not sure how to use `task_prerun` without adding a decorator everywhere

[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/tasks/base.py:110]: I'm not sure if `after_task_publish` will work. It's fired after the task has already been sent. At that point it's too late to modify the messgage.

https://docs.celeryq.dev/en/latest/userguide/signals.html#before-task-publish might work, although I'm not totally sure what injecting/fetching the parameters would look like there

[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/tasks/base.py:110]: @wedamija @NisanthanNanthakumar I don't see a way to move forward with signals. I need to inject the start time into the task in order to determine the duration. And if I inject the start time, I have to pop it off before we call the underlying function. Not sure why this current approach won't work.

File: src/sentry/celery.py [hunk 2]
```diff
@@ -83,6 +84,10 @@ def good_use_of_pickle_or_bad_use_of_pickle(task, args, kwargs):
 class SentryTask(Task):
     Request = "sentry.celery:SentryRequest"
 
+    def delay(self, *args, **kwargs):
+        kwargs["__start_time"] = datetime.now().timestamp() * 1000
+        return super().delay(*args, **kwargs)
+
     def apply_async(self, *args, **kwargs):
```
Comment (wedamija) [inline src/sentry/celery.py:91]: Probably should do this in `apply_async` as well

File: src/sentry/celery.py [hunk 3]
```diff
@@ -83,6 +84,10 @@ def good_use_of_pickle_or_bad_use_of_pickle(task, args, kwargs):
 class SentryTask(Task):
     Request = "sentry.celery:SentryRequest"
 
+    def delay(self, *args, **kwargs):
+        kwargs["__start_time"] = datetime.now().timestamp() * 1000
```
Comment (wedamija) [inline src/sentry/celery.py:88]: Nit: Should we just do the `* 1000` when calculating the duration so it's not in two places?

**Review by wedamija — COMMENTED** [review]
Looks like this broke a bunch of tests

**Comment by scefali** [comment]
@wedamija I'll try to fix if I can find some time...

File: src/sentry/celery.py (hunk 2 above)
Comment (scefali) [reply src/sentry/celery.py:91]: @wedamija Actually, are we already recording the timing with `with metrics.timer("jobs.delay", instance=self.name):`?

Comment (wedamija) [reply src/sentry/celery.py:91]: No, that's just timing how long it takes to send the task to celery afaict

Comment (scefali) [reply src/sentry/celery.py:91]: @wedamija hmmm I'm not sure how to get this to work for `apply_async` since we call `Task.apply_async` directly.

**Review by wedamija — APPROVED** [review]

**Pull request merged**